k2/csrc/fsa_utils.h

```diff
@@ -0,0 +1,85 @@
+/**
+ * @brief Utilities for reading, writing and creating FSAs.
+ *
+ * @copyright
+ * Copyright (c)  2020  Mobvoi Inc.        (authors: Fangjun Kuang)
+ *
+ * @copyright
+ * See LICENSE for clarification regarding multiple authors
+ */
+
+#ifndef K2_CSRC_FSA_UTILS_H_
+#define K2_CSRC_FSA_UTILS_H_
+
+#include <string>
+
+#include "k2/csrc/array.h"
+#include "k2/csrc/fsa.h"
+
+namespace k2 {
+
+/*
+  Create an Fsa from string.
+
+  The string `s` consists of lines. Every line, except the line for
+  the final state, has one of the following two formats:
+
+  (1)
+      src_state dest_state label cost
+  which means the string represents an acceptor.
+
+  (2)
+      src_state dest_state label aux_label cost
+  which indicates the string is a transducer.
+
+  The line for the final state has the following format:
+
+      final_state
+
+  CAUTION: The line for the final state contains NO cost.
+  When an arc's dest state is the final state, we put the cost
+  on the arc and set its label to -1.
+
+
+  @param [in]   s   The input string. See the above description for its format.
+  @param [in]   negate_scores
+                    If true, the string form has the weights as costs,
+                    not scores, so we negate as we read.
+  @param [out]  aux_labels
+                    Used only when it is a transducer. It contains the
+                    aux_label of each arc.
+
+  @return It returns an Fsa.
+ */
+Fsa FsaFromString(const std::string &s, bool negate_scores = false,
```

Comment:
What's more important, to me, than reading/writing to files, is printing to a string in the same format
that FsaFromString uses.  I think that's more likely to be used in practice.

My plan for file-based I/O was to not really do this from C++ but instead do it from Python via vectors of arcs,
i.e. write the tensor to disk using torch or TF mechanisms according to the frontend framework.  I'm open
to discussion on this though.
I'd rather not confuse matters by having un-used APIs around though, if we stick to my original plan.

Reply:
Removed from c++.